### PR TITLE
add annotation for private network

### PR DESCRIPTION
**Pull request by deitch**
In addition to the BGP annotations we already use. This came as an outgrowth of Gardener work, but I started seeing other potential use cases. In any case, the right place to do this is CPEM.

@displague I know this will cause you to have to rebase your big PR; sorry about that. this was getting in the way of some other work.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -208,6 +208,7 @@ This section lists each configuration option, and whether it can be set by each
 | Kubernetes annotation to set BGP peer's IPs |   | `METAL_ANNOTATION_PEER_IPS` | `annotationPeerIPs` | `"metal.equinix.com/peer-ip"` |
 | Kubernetes annotation to set source IP for BGP peering |   | `METAL_ANNOTATION_SRC_IP` | `annotationSrcIP` | `"metal.equinix.com/src-ip"` |
 | Kubernetes annotation to set BGP MD5 password, base64-encoded (see security warning below) |   | `METAL_ANNOTATION_BGP_PASS` | `annotationBGPPass` | `"metal.equinix.com/bgp-pass"` |
+| Kubernetes annotation to set the CIDR for the network range of the private address |  | `METAL_ANNOTATION_NETWORK_IPV4_ PRIVATE` |  `annotationNetworkIPv4Private` | `metal.equinix.com/network/4/private` |
```
Comment (displague) [inline README.md:211]:
```suggestion
| Kubernetes annotation to set the CIDR for the network range of the private address |  | `METAL_ANNOTATION_NETWORK_IPV4_PRIVATE` |  `annotationNetworkIPv4Private` | `metal.equinix.com/network/4/private` |
```

Comment (displague) [reply README.md:211]: Are these "private" VLAN addresses or 10/8 project-scoped management addresses?

I realize the terms are overlapping and fuzzy:
https://metal.equinix.com/developers/docs/networking/ip-addresses/

Comment (deitch) [reply README.md]: 10/8 project-scoped. What k8s also calls "private" or "internal"

**Comment by deitch** [comment]
Took your doc suggestion. Anything else @displague ?

File: main.go [hunk 2]
```diff
@@ -191,6 +192,12 @@ func getMetalConfig(providerConfig string) (metal.Config, error) {
 		config.AnnotationBGPPass = annotationBGPPass
 	}
 
+	config.AnnotationNetworkIPv4Private = metal.DefaultAnnotationNetworkIPv4Private
+	annotationNetworkIPv4Private := os.Getenv(envVarAnnotationIPv4NetworkPrivate)
```
Comment (displague) [inline main.go:196]: perhaps the env var constant could match the config name: envVarAnnotationNetworkIPv4Private (to make future code searches easier)

**Review by displague — APPROVED** [review]
This looks reasonable.  I pointed out a logger typo.

I wonder about the intention of this PR. It seems that this allows for a user-specified private range, by setting the annotation and value rather than looking one up. This would better accommodate VLAN configured networks.  Some documentation describing that use case (and how to accomplish it) may be helpful.

How is this annotation used?

How is this value different than the node InternalIP address?  Why are we not reflecting that address in this annotation?

File: metal/devices.go [hunk 3]
```diff
@@ -252,3 +256,84 @@ func (i *instances) deviceFromProviderID(providerID string) (*packngo.Device, er
...
+					}
+				} else {
+					klog.V(2).Infof("instances.reconcileNodes(): no change to annotations for %s", node.Name)
+				}
+			}
+		}
+	case ModeRemove:
+		klog.V(2).Info("instances.reconcileNodes(): nothing to do for removing nodes")
+	}
+	klog.V(2).Info("instamces.reconcileNodes(): complete")
```
[the remark this answers is not in the document]
Comment (deitch) [reply metal/devices.go:316]: Nah, you never heard of `instamces`? AWS has ec2 "instances", and EQXM has "instamces", ("m" for "metal") 😆

[the remark this answers is not in the document]
Comment (deitch) [reply metal/devices.go]: Fixed

File: main.go (hunk 2 above)
Comment (deitch) [reply main.go:196]: That's a good idea. Will do.

Comment (deitch) [reply main.go]: Done.

**Comment by deitch** [comment]
>  I wonder about the intention of this PR. It seems that this allows for a user-specified private range, by setting the annotation and value rather than looking one up from the EM API. This would better accommodate VLAN configured networks where the EM API is unaware of the IP configuration within the VLAN. Some documentation describing that use case (and how to accomplish it) may be helpful.

That wasn't at all the intent here, see below. That having been said, I am open to new intents.

> How is this annotation used?

This annotation - like the others - convey information about the device that normally would be gotten via metadata (if on the node) or via the EQXM API. Actually, CPEM uses the EQXM API to get these and add them.

> How is this value different than the node InternalIP address? Why are we not reflecting that address (and range) in this annotation?

The internalIP address is the actual address of the node itself. This annotation gets the range in which the internalIP sits. kubelet already gets the private IP and adds it as a property of the node, so we don't need to annotate it.

Here is an example of a node which I currently am running:

```json
      {
        "id": "eb8d4b25-9780-4b53-a0b5-5d1ed99cca48",
        "address_family": 4,
        "netmask": "255.255.255.254",
        "created_at": "2021-06-14T14:37:48Z",
        "public": false,
        "cidr": 31,
        "management": true,
        "enabled": true,
        "network": "10.99.173.130",
        "address": "10.99.173.131",
        "gateway": "10.99.173.130",
        "parent_block": {
          "network": "10.99.173.128",
          "netmask": "255.255.255.128",
          "cidr": 25,
          "href": "/ips/e9753813-eec6-4b4a-8db9-5712fbbeda21"
        }
      }
```

When this node is created, the node's InternalIP will be `10.99.173.131/31`. But what private network range is it in? That is given by the `parent_block` (same on the API), which is `10.99.173.128/25`.

The annotation adds that.

Does that explain it?

**Comment by displague** [comment]
> The internalIP address is the actual address of the node itself. This annotation gets the range in which the internalIP sits. kubelet already gets the private IP and adds it as a property of the node, so we don't need to annotate it.

These are not necessarily the same networks, which is why I think it would help to expand on the intent and usage.  The `InternalIP` could be a 192.168/16 member configured on a VLAN while the EM API reports a 10/25 network.

**Comment by deitch** [comment]
Hmm that is a good point. 

The intended usage is where you need to know what the network range is. With other clouds, you actually know it because you set up a VPC and subnets. Here, that doesn't exist. 

What are you suggesting?

**Comment by displague** [comment]
I've been wondering about what CPEM might look like in pure layer-2 clusters. Cloud-provide-kube-vip may be a better fit in that scenario. In a mixed cluster, with control plane nodes in hybrid mode and worker nodes in layer-2 we may have need for varied internal network configurations. 

The upcoming gateway feature opens up some configurations I'd like to explore.

**Comment by deitch** [comment]
> I've been wondering about what CPEM might look like in pure layer-2 clusters. Cloud-provide-kube-vip may be a better fit in that scenario. In a mixed cluster, with control plane nodes in hybrid mode and worker nodes in layer-2 we may have need for varied internal network configurations.
>
> The upcoming gateway feature opens up some configurations I'd like to explore.

This does open up some interesting new directions. I will hit you up on slack, we can arrange some time to set up a discussion, including others who know. 

In the meantime, since that will be a material rework, can we move ahead with this one?

**Review by displague — APPROVED** [review]

**Pull request merged**